rfd/0028-cluster-config-resources.md

```diff
@@ -0,0 +1,288 @@
+    AuditConfig Audit;
+
+    // ClientIdleTimeout sets global cluster default setting for client idle
+    // timeouts
+    int64 ClientIdleTimeout;
+
+    // DisconnectExpiredCert provides disconnect expired certificate setting -
+    // if true, connections with expired client certificates will get disconnected
+    bool DisconnectExpiredCert;
+
+    // KeepAliveInterval is the interval the server sends keep-alive messsages
+    // to the client at.
+    int64 KeepAliveInterval;
+
+    // KeepAliveCountMax is the number of keep-alive messages that can be missed
+    // before
+    // the server disconnects the connection to the client.
+    int64 KeepAliveCountMax;
+
+    // LocalAuth is true if local authentication is enabled.
+    bool LocalAuth;
+
+    // SessionControlTimeout is the session control lease expiry and defines
+    // the upper limit of how long a node may be out of contact with the auth
+    // server before it begins terminating controlled sessions.
+    int64 SessionControlTimeout;
+}
+```
+
+It is proposed to distribute them in the following fashion:
+
+```proto
+// ClusterID field is moved into ClusterName resource.
+message ClusterName {
+    ...
+    string ClusterID;
+}
+
+message SessionRecordingConfig {
+    // ClusterConfig.SessionRecording is renamed to SessionRecordingConfig.Mode.
+    string Mode;
+    bool ProxyChecksHostKeys;
+}
+
+message ClusterNetworkingConfig {
+    int64 ClientIdleTimeout;
+    int64 KeepAliveInterval;
+    int64 KeepAliveCountMax;
+    int64 SessionControlTimeout;
+}
+
+// The already existing AuditConfig is turned into a standalone resource.
+message AuditConfig { ... }
+
+// DisconnectExpiredCert & LocalAuth fields are moved into ClusterAuthPreference.
+message ClusterAuthPreference {
+    ...
+    bool AllowLocalAuth;
+    bool DisconnectExpiredCert;
+}
+```
+
+### Restricting to a subset of values of a field
+
+Some users should not be capable of setting a field to any of the otherwise
+accepted values.  It is proposed to satisfy this requirement by introducing new
+RBAC verbs, interpreted as implying additional constraints. Such new verbs
+would be evaluated together with other RBAC verbs in the auth API handlers.
+
+In particular, this is needed for `SessionRecordingConfig` which has two
+fields:
+* `mode` with the possible values: `off`, `node`, `node-sync`,
+  `proxy`, `proxy-sync`;
+* a boolean `proxy_checks_host_keys`.
+
+Some users should be subject to the following restrictions:
+1. The value of `mode` can be changed only if its original value
+   is one of `off`, `node` or `node-sync`.
+1. The value of `mode` can be changed to only one of `off`, `node`
+   or `node-sync`.
+1. The value of `proxy_checks_host_keys` cannot be changed.
+
+This would be achieved by a new RBAC verb `update.no_proxy_recording`. A user could then
+be assigned a role featuring `update.no_proxy_recording` instead of the regular, more
+permissive `update`:
+```
+  allow:
```

Comment:
I think it's an interesting concept, however this won't work unfortunately for cloud. For cloud you'd need to have some fields that can not be changed even by cluster administrator.

I propose we remove this change then out of scope of the RFD and instead use lib/modules to restrict some values that we know can not be changed by the cloud:

https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L66

```
interface Modules {
   ValidateResource(res Resource) error
}
```

Then cloud build would implement 

```go
ValidateResource(res Resource) error {
   switch r := res.(type) {
        case *ClusterNetworkingConfig:
            if features.Cloud && r.GetMetadata().GetLabels().Source == "user" // whatever we defined for user {
                  return trace.BadParameter("cloud users not modify the networking parameters")
            }
       case *AuthConfig:
           if r.GetSecondFactor() == "off" {
               return trace.BadParameter("cloud can not have second factor off")
           }
}
}
```

Reply:
Would it be OK to let this RFD deal only with the `ClusterConfig` reorganization and entirely omit this discussion of "Restricting to a subset of values of a field"?

Reply:
@andrejtokarcik let's just add a short section explaining _how_ Cloud will restrict modifications (hardcoded checks via https://github.com/gravitational/teleport/blob/master/lib/modules/modules.go#L64-L75) without describing any specific limitations (e.g. "can't turn of 2FA").

We can let the Cloud team figure out what they want to restrict outside of this RFD.

Reply:
Updated.